### PR TITLE
recognize equivalent circular data structures in ‘equals’

### PR DESCRIPTION
**Pull request by davidchambers**
Many thanks to @vyznev for suggesting this approach on [Stack Overflow][1].

@Avaq, this is the part where you try to break it. :)

[1]: http://stackoverflow.com/a/40622794/312785

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -602,7 +602,7 @@
   var Array$prototype$equals = function(other) {
     if (other.length !== this.length) return false;
     for (var idx = 0; idx < this.length; idx += 1) {
-      if (!equals(other[idx], this[idx])) return false;
+      if (!equals(this[idx], other[idx])) return false;
```
Comment (davidchambers) [inline index.js]: In order to compare `[1, <Circular>]` and `[1, [1, <Circular>]]` we expect to compare `<Circular>` and `[1, <Circular>]` (after comparing `1` and `1`), but with the previous argument order the values would be flipped: `[1, <Circular>]` and `<Circular>`. The change above isn't necessary, but avoids confusion when debugging.

File: index.js [hunk 2]
```diff
@@ -738,6 +738,11 @@
     return function(_) { return x; };
   };
 
+  //  Function$prototype$equals :: Function ~> Function -> Boolean
+  var Function$prototype$equals = function(other) {
+    return other === this;
+  };
+
```
Comment (davidchambers) [inline index.js]: Having removed `Object(x) === Object(y)` from `equals` we need to define `Function#fantasy-land/equals`.

File: test/index.js [hunk 3]
```diff
@@ -383,7 +386,8 @@ test('toString', function() {
   eq(Z.toString(['foo', 'bar']), '["foo", "bar"]');
   eq(Z.toString(/x/.exec('xyz')), '["x", "index": 0, "input": "xyz"]');
   eq(Z.toString((function() { var xs = []; xs.z = true; xs.a = true; return xs; }())), '["a": true, "z": true]');
-  eq(Z.toString(circular), '["foo", <Circular>]');
+  eq(Z.toString(ones), '[1, <Circular>]');
+  eq(Z.toString(ones_), '[1, [1, <Circular>]]');
```
Comment (davidchambers) [inline test/index.js]: Having two string representations of one value is somewhat disturbing, but I don't see an alternative.

File: index.js [hunk 4]
```diff
@@ -932,25 +944,27 @@
...
-    //  $seen :: Array Any
-    var $seen = [];
+    //  $pairs :: Array (Pair Any Any)
+    var $pairs = [];
 
-    //  equal :: (a, b) -> Boolean
-    var equal = function(x, y) {
-      $seen.push(x, y);
+    return function equals(x, y) {
+      if (type(x) !== type(y)) {
```
Comment (rjmk) [inline index.js]: Incredible how much of the solution Sanctuary's typechecking captures!

Comment (davidchambers) [inline index.js]: I :heart: the simplicity of the approach. _Are you two members of the same type? Is that type a setoid? Okay, we'll let its `fantasy-land/equals` method determine whether you are equal._

**Pull request merged**